### PR TITLE
[#604] Add run-ct-test code lens, refactor code lens behaviour

### LINKED ISSUE #604: Implement Code Lens to run CT test

### PR DESCRIPTION
**Pull request by robertoaloi**
### Description

This PR introduces a new _Code Lens_ and a _Command_, named "ct-run-test", through which is now possible to run Common Test testcases by clicking on a button from a test suite.

The code lens is activated automatically whenever the `ct.hrl` header is included in a module. The lens will be enabled for all functions of arity `1` in the test suite and a _blacklist_ avoids functions such as `init_per_suite/1` or `end_per_suite/` to be included.

The CT test is executed as a _background job_ and a CT hook is used to report diagnostics in case of failed or skipped tests.

I used this opportunity to refine the _code lens_ behaviour, introducing some extra callbacks that should hopefully simplify the creation of new code lenses. I will document the behaviour properly once the API is settled.

More things need to be implemented or fixed, but I believe this PR can give us a starting point.

Finally, I moved some more type definitions away from the `erlang_ls.hrl`. Having them as part of individual modules give us a nicer way to interact with types (e.g. via type completion) and avoids collisions between names. I plan to go in this direction for the whole project.

Fixes #604 .

### CODE REVIEW

**Comment by alanz** [comment]
I notice it currently drops `.beam` files into the test directory next to the source files.

File: src/els_cth.erl [hunk 1]
```diff
@@ -0,0 +1,37 @@
+-module(els_cth).
```
Comment (alanz) [inline src/els_cth.erl]: This is quite a cryptic name, should there perhaps be a document comment saying what this module is for?

**Comment by robertoaloi** [comment]
> I notice it currently drops `.beam` files into the test directory next to the source files.

This is probably due to the `auto_compile` option. It's one of the things I need to follow-up. Should we create a ticket for that and address it separately?

**Comment by robertoaloi** [comment]
> But I presume this is a work in progress.

It is indeed. There are many scenarios where things do not work correctly. My plan is to start having the feature (which per se should be quite stable, even if the tests could fail for various reasons) so that people can play around with it and report specific issues, that we can address individually. If we see that the lens creates too much confusion and we haven't address all the major concerns, we can disable it by default before next release.

@alanz Can I ask you to report the specific issues you noticed?

File: src/els_cth.erl (hunk 1 above)
Comment (robertoaloi) [reply src/els_cth.erl]: 100% agree. Do you have a suggestion?

File: src/els_command_ct_run_test.erl [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+-spec execute(map()) -> ok.
+execute(#{ <<"module">>   := M
+         , <<"function">> := F
+         , <<"arity">>    := A
+         , <<"line">>     := Line
+         , <<"uri">>      := Uri
+         }) ->
+  Msg = io_lib:format("Running Common Test [mfa=~s:~s/~p]", [M, F, A]),
+  lager:info(Msg, []),
```
Comment (jfacorro) [inline src/els_command_ct_run_test.erl]: Why this combination of `io_lib:format/2` + `lager:info/2` and not only use `lager:info/2`?
```suggestion
  lager:info("Running Common Test [mfa=~s:~s/~p]", [M, F, A]),
```

File: src/els_cth.erl [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  State.
+
+-spec on_tc_skip(suite(), testcase(), skip_reason(), state()) -> state().
+on_tc_skip(_Suite, _TestCase, {_ReasonType, Reason}, State) ->
+  publish_result(Reason, State),
+  State.
+
+-spec publish_result(reason(), state()) -> ok.
+publish_result(Reason, #{options := #{uri := Uri, line := Line}}) ->
+  Message = unicode:characters_to_binary(io_lib:format("~p", [Reason])),
```
Comment (jfacorro) [inline src/els_cth.erl]:
```suggestion
  Message = els_utils:to_binary(io_lib:format("~p", [Reason])),
```

File: src/els_code_lens_ct_run_test.erl [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+-spec pois(els_dt_document:item()) -> [poi()].
+pois(Document) ->
+  Functions = els_dt_document:pois(Document, [function]),
+  [POI || #{id := {F, 1}} = POI <- Functions, not is_blacklisted(F)].
+
+-spec precondition(els_dt_document:item()) -> boolean().
+precondition(Document) ->
+  Includes = els_dt_document:pois(Document, [include_lib]),
+  case [POI || #{id := "common_test/include/ct.hrl"} = POI <- Includes] of
```
Comment (jfacorro) [inline src/els_code_lens_ct_run_test.erl]: Does this mean that we are detecting if a module is a Common Test suite based on it including this header file?

File: src/els_code_lens_ct_run_test.erl [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+title(_POI) ->
+  <<"Run test">>.
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+-spec is_blacklisted(atom()) -> boolean().
+is_blacklisted(Function) ->
+  lists:member(Function, [init_per_suite, end_per_suite]).
```
Comment (jfacorro) [inline src/els_code_lens_ct_run_test.erl]: Why not include all the other callbacks from Common Test?

File: src/els_command_ct_run_test.erl (hunk 2 above)
Comment (robertoaloi) [reply src/els_command_ct_run_test.erl]: Because the `Message` is re-used in the `Title`.

File: src/els_code_lens_ct_run_test.erl (hunk 4 above)
Comment (robertoaloi) [reply src/els_code_lens_ct_run_test.erl]: Exactly. We can refine this strategy in the future if we want to (e.g. if the name ends up with a specific suffix, etc).

File: src/els_code_lens_ct_run_test.erl (hunk 5 above)
Comment (robertoaloi) [reply src/els_code_lens_ct_run_test.erl]: Because the other ones do not have arity `1`. But I missed `group`, which I'm adding now.

**Comment by robertoaloi** [comment]
> I notice it currently drops `.beam` files into the test directory next to the source files.

Created https://github.com/erlang-ls/erlang_ls/issues/610 as a follow-up

**Pull request merged**